From 1baf89d9e6c581359dc179684bccf0c3efbb4f9b Mon Sep 17 00:00:00 2001
From: Michal Hrusecky <michal.hrusecky@turris.com>
Date: Tue, 20 Jul 2021 10:01:00 +0200
Subject: [PATCH] snort3: Provide more information in lua bindings

Making more information available in lua bindings, patches sent
upstream.

https://github.com/snort3/snort3/pull/180
---
 net/snort3/Makefile                           |  2 +-
 ..._luajit-Make-MAC-addresses-available.patch | 68 +++++++++++++++++++
 ...ake-destination-IP-address-available.patch | 59 ++++++++++++++++
 3 files changed, 128 insertions(+), 1 deletion(-)
 create mode 100644 net/snort3/patches/0001-alert_luajit-Make-MAC-addresses-available.patch
 create mode 100644 net/snort3/patches/0002-alert_luajit-Make-destination-IP-address-available.patch

diff --git a/net/snort3/Makefile b/net/snort3/Makefile
index 4530596a8..fee8f6569 100644
--- a/net/snort3/Makefile
+++ b/net/snort3/Makefile
@@ -7,7 +7,7 @@ include $(TOPDIR)/rules.mk
 
 PKG_NAME:=snort3
 PKG_VERSION:=3.1.0.0
-PKG_RELEASE:=3
+PKG_RELEASE:=4
 
 PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
 PKG_SOURCE_URL:=https://www.snort.org/downloads/snortplus/
diff --git a/net/snort3/patches/0001-alert_luajit-Make-MAC-addresses-available.patch b/net/snort3/patches/0001-alert_luajit-Make-MAC-addresses-available.patch
new file mode 100644
index 000000000..5e7812f76
--- /dev/null
+++ b/net/snort3/patches/0001-alert_luajit-Make-MAC-addresses-available.patch
@@ -0,0 +1,68 @@
+From 24e2c411189f77224846571469aa6c010e3bf0dc Mon Sep 17 00:00:00 2001
+From: Michal Hrusecky <michal.hrusecky@turris.com>
+Date: Thu, 6 May 2021 22:25:03 +0200
+Subject: [PATCH 1/2] alert_luajit: Make MAC addresses available
+
+Add ether_src and ether_dst to the packet information in LUA to allow
+better alerting.
+---
+ src/loggers/alert_luajit.cc    | 22 ++++++++++++++++++++++
+ src/managers/lua_plugin_defs.h |  2 ++
+ 2 files changed, 24 insertions(+)
+
+diff --git a/src/loggers/alert_luajit.cc b/src/loggers/alert_luajit.cc
+index a4ad98d49..52ac081e8 100644
+--- a/src/loggers/alert_luajit.cc
++++ b/src/loggers/alert_luajit.cc
+@@ -21,6 +21,7 @@
+ #include "config.h"
+ #endif
+ 
++#include "protocols/eth.h"
+ #include "detection/ips_context.h"
+ #include "detection/signature.h"
+ #include "events/event.h"
+@@ -82,6 +83,27 @@ SO_PUBLIC const SnortPacket* get_packet()
+     lua_packet.sp = packet->ptrs.sp;
+     lua_packet.dp = packet->ptrs.dp;
+ 
++    if ( !(packet->proto_bits & PROTO_BIT__ETH) ) {
++        lua_packet.ether_dst = lua_packet.ether_src = "";
++        return &lua_packet;
++    }
++
++    static char eth_src[20], eth_dst[20];
++
++    const eth::EtherHdr* eh = layer::get_eth_layer( packet );
++
++    snprintf(eth_src, 20, "%02X:%02X:%02X:%02X:%02X:%02X", eh->ether_src[0],
++        eh->ether_src[1], eh->ether_src[2], eh->ether_src[3],
++        eh->ether_src[4], eh->ether_src[5]);
++
++    lua_packet.ether_src = (const char*)eth_src;
++
++    snprintf(eth_dst, 20, "%02X:%02X:%02X:%02X:%02X:%02X", eh->ether_dst[0],
++        eh->ether_dst[1], eh->ether_dst[2], eh->ether_dst[3],
++        eh->ether_dst[4], eh->ether_dst[5]);
++
++    lua_packet.ether_dst = (const char*)eth_dst;
++
+     return &lua_packet;
+ }
+ 
+diff --git a/src/managers/lua_plugin_defs.h b/src/managers/lua_plugin_defs.h
+index 4efc683dd..5fce3682e 100644
+--- a/src/managers/lua_plugin_defs.h
++++ b/src/managers/lua_plugin_defs.h
+@@ -55,6 +55,8 @@ struct SnortPacket
+     uint64_t num;
+     unsigned sp;
+     unsigned dp;
++    const char* ether_src;
++    const char* ether_dst;
+ };
+ 
+ extern "C"
+-- 
+2.31.1
+
diff --git a/net/snort3/patches/0002-alert_luajit-Make-destination-IP-address-available.patch b/net/snort3/patches/0002-alert_luajit-Make-destination-IP-address-available.patch
new file mode 100644
index 000000000..2a2dee652
--- /dev/null
+++ b/net/snort3/patches/0002-alert_luajit-Make-destination-IP-address-available.patch
@@ -0,0 +1,59 @@
+From 29bd324a313be710aa6dd2a929c71a95b4ccb685 Mon Sep 17 00:00:00 2001
+From: Michal Hrusecky <michal.hrusecky@turris.com>
+Date: Thu, 6 May 2021 22:37:56 +0200
+Subject: [PATCH 2/2] alert_luajit: Make destination IP address available
+
+Make sure we have destination IP address available at hand in lua
+alerting modules.
+---
+ src/loggers/alert_luajit.cc    | 7 +++++++
+ src/managers/lua_plugin_defs.h | 2 +-
+ 2 files changed, 8 insertions(+), 1 deletion(-)
+
+diff --git a/src/loggers/alert_luajit.cc b/src/loggers/alert_luajit.cc
+index 52ac081e8..c1edd7d24 100644
+--- a/src/loggers/alert_luajit.cc
++++ b/src/loggers/alert_luajit.cc
+@@ -82,6 +82,7 @@ SO_PUBLIC const SnortPacket* get_packet()
+     lua_packet.num = packet->context->packet_number;
+     lua_packet.sp = packet->ptrs.sp;
+     lua_packet.dp = packet->ptrs.dp;
++    lua_packet.dst_addr = "";
+ 
+     if ( !(packet->proto_bits & PROTO_BIT__ETH) ) {
+         lua_packet.ether_dst = lua_packet.ether_src = "";
+@@ -104,6 +105,12 @@ SO_PUBLIC const SnortPacket* get_packet()
+ 
+     lua_packet.ether_dst = (const char*)eth_dst;
+ 
++    if ( packet->has_ip() or packet->is_data() ) {
++        static char ip[50];
++        packet->ptrs.ip_api.get_dst()->ntop(ip, 50);
++        lua_packet.dst_addr = ip;
++    }
++
+     return &lua_packet;
+ }
+ 
+diff --git a/src/managers/lua_plugin_defs.h b/src/managers/lua_plugin_defs.h
+index 5fce3682e..7de831f81 100644
+--- a/src/managers/lua_plugin_defs.h
++++ b/src/managers/lua_plugin_defs.h
+@@ -50,13 +50,13 @@ const struct SnortEvent* get_event();
+ 
+ struct SnortPacket
+ {
+-    // FIXIT-L add ip addrs and other useful foo to lua packet
+     const char* type;
+     uint64_t num;
+     unsigned sp;
+     unsigned dp;
+     const char* ether_src;
+     const char* ether_dst;
++    const char* dst_addr;
+ };
+ 
+ extern "C"
+-- 
+2.31.1
+
-- 
2.32.0

